src/Cli/dotnet/ToolPackage/ToolPackageDownloader.cs

```diff
@@ -0,0 +1,311 @@
+                {
+                    throw new ToolPackageException(
+                        string.Format(
+                            CommonLocalizableStrings.ToolPackageConflictPackageId,
+                            packageId,
+                            version.ToNormalizedString()));
+                }
+
+                if (Directory.Exists(packagePath))
+                {
+                    throw new ToolPackageException(
+                        string.Format(
+                            CommonLocalizableStrings.ToolPackageConflictPackageId,
+                            packageId,
+                            version.ToNormalizedString()));
+                }
+
+                var packageHash = Convert.ToBase64String(new CryptoHashProvider("SHA512").CalculateHash(reader.GetNuspec()));
+                var hashPath = new VersionFolderPathResolver(hashPathLocation).GetHashPath(packageId.ToString(), version);
+
+                Directory.CreateDirectory(Path.GetDirectoryName(hashPath));
+                File.WriteAllText(hashPath, packageHash);
+            }
+
+            // Extract the package
+            var nupkgDir = Path.Combine(hashPathLocation, packageId.ToString(), version.ToString());
+            var filesInPackage = await _nugetPackageDownloader.ExtractPackageAsync(packagePath, new DirectoryPath(nupkgDir));
+
+            return version;
+        }
+
+        private static void CreateAssetFiles(
+            PackageId packageId,
+            NuGetVersion version,
+            DirectoryPath nugetLocalRepository,
+            DirectoryPath assetFileDirectory,
+            string runtimeJsonGraph)
+        {
+            // To get runtimeGraph:
+            var runtimeGraph = JsonRuntimeFormat.ReadRuntimeGraph(runtimeJsonGraph);
+
+            // Create ManagedCodeConventions:
+            var conventions = new ManagedCodeConventions(runtimeGraph);
+
+            //  Create LockFileTargetLibrary
+            var lockFileLib = new LockFileTargetLibrary()
+            {
+                Name = packageId.ToString(),
+                Version = version,
+                Type = LibraryType.Package,
+                PackageType = new List<PackageType>() { PackageType.DotnetTool }
+            };
+
+            //  Create NuGetv3LocalRepository
+            NuGetv3LocalRepository localRepository = new(nugetLocalRepository.Value);
+            var package = localRepository.FindPackage(packageId.ToString(), version);
+
+            var collection = new ContentItemCollection();
+            collection.Load(package.Files);
+
+            //  Create criteria
+            var managedCriteria = new List<SelectionCriteria>(1);
+            // Use major.minor version of currently running version of .NET
+            var currentTargetFramework = new NuGetFramework(FrameworkConstants.FrameworkIdentifiers.NetCoreApp, new Version(Environment.Version.Major, Environment.Version.Minor));
```

Comment:
We should take the target framework to use as a parameter to this method.  If it's not set then we should set it to the default value.